File: backend/strategy-game-common-data/src/main/kotlin/io/github/smiley4/strategygame/backend/commondata/Command.kt
package io.github.smiley4.strategygame.backend.commondata


class Command<T : CommandData>(
    val id: Id,
    val game: Game.Id,
    val user: User.Id,
    val turn: Int,
    val data: T
) {

    @JvmInline
    value class Id(val value: String)

}

sealed class CommandData {

    /**
     * Move the given world object along the given path.
     */
    class Move(
        val worldObject: WorldObject.Id,
        val path: List<Tile.Ref>
    ) : CommandData()


    /**
     * Disband / Delete the given world object.
     */
    class Disband(
        val worldObject: WorldObject.Id,
    ) : CommandData()


    /**
     * Use the given world object to construct a tile improvement.
     */
    class ConstructTileImprovement(
        val worldObject: WorldObject.Id,
        val improvement: TileImprovementType,
    ) : CommandData()

    /**
     * Use the given world object to spawn a new settlement.
     */
    class SpawnSettlement(
        val worldObject: WorldObject.Id,
        val tile: Tile.Ref,
        val settlementName: String
    ) : CommandData()

}

File: frontend/js/src/models/command/command.ts
import {TileSummary} from "../tile/tileSummary";
import {BrandedId} from "../../common/brandedId";
import {WorldObject} from "../worldobject/worldObject";
import {UID} from "../../common/uid";

/**
 * A command given by the player in the current turn.
 */
export type Command = Command.Move
	| Command.Disband
	| Command.ConstructTileImprovement
	| Command.CreateSettlement

export namespace Command {

	export type Id = BrandedId<string, "CommandId">;

	export function genId(): Id {
		return UID.generate() as Id;
	}

	export enum Type {
		Move = "move",
		Disband = "disband",
		ConstructTileImprovement = "construct-tile-improvement",
		CreateSettlement = "create-settlement"
	}

	export type Mapping = {
		[Type.Move]: Move,
		[Type.Disband]: Disband,
		[Type.ConstructTileImprovement]: ConstructTileImprovement
		[Type.CreateSettlement]: CreateSettlement
	}

	interface BaseCommand {
		type: Type,
		id: Id,
	}

	export interface Move extends BaseCommand {
		type: Type.Move;
		worldObjectId: WorldObject.Id
		path: TileSummary[],
	}


	export interface Disband extends BaseCommand {
		type: Type.Disband;
		worldObjectId: WorldObject.Id;
	}

	export interface ConstructTileImprovement extends BaseCommand {
		type: Type.ConstructTileImprovement;
		worldObjectId: WorldObject.Id;
		tileImprovementType: string;
	}

	export interface CreateSettlement extends BaseCommand {
		type: Type.CreateSettlement;
		worldObjectId: WorldObject.Id;
		tile: TileSummary;
		name: string;
	}

}